File: Mars.py
#!/usr/bin/env python3
import numpy as np
class Heap:
    """
    Max heap using distance as key. Here we can think of terrain
    as edges weight, so we can find the path with maximum length.
    """
    def __init__( self, dist ):
        self.key = dist         # reference (numpy array)
        self.arr = [( -1, -1 )] # 1-based heap
        for vtx, value in np.ndenumerate( dist ):
            # skip rough terrain
            if value != 1: self.insert( vtx )

    def __str__( self ):
        if self.empty(): return "The Heap is empty!"
        levl = 1 # 1,3,7,15,...
        buff = ""
        back = len( self.arr ) - 1 # Last Mohican
        j = 0
        while j < back:
            j += 1
            v = self.arr[j]
            k = self.key[v]
            buff += "({},{})[{}]".format( *v, k )
            if j == levl or j == back:
                buff += '\n'
                levl = ( levl << 1 ) + 1
        return buff

    def swap( self, i, j ):
        self.arr[i], self.arr[j] = self.arr[j], self.arr[i]

    def lt( self, i, j ):
        return (self.key[self.arr[i]] < self.key[self.arr[j]])

    def bubbleup( self, j ):
        while True:
            p = j // 2 # parent node
            if p == 0: break
            if self.lt( p, j ): # Does parent vtx have less dist?
                self.swap( p, j )
                j = p
            else: break

    def insert( self, vtx ):
        self.arr.append( vtx )
        self.bubbleup( len( self.arr ) - 1 )

    def find( self, vtx ):
        for j in range( 1, len( self.arr )):
            if vtx == self.arr[j]: return j
        return -1

    def update( self, vtx ):
        """
        Here vtx dist has changed, and we have to update the heap.
        """
        j = self.find( vtx )
        self.bubbleup( j )

    def maxchild( self, j ):
        l = j << 1              # left child (2j)
        n = len( self.arr ) - 1 # last index
        if l > n:
            return -1
        if l == n:
            return l
        else:
            r = l + 1           # right child
            if self.lt( l, r ): # left < right?
                return r
            else:
                return l
        
    def pop( self ):
        """ Ck if empty before popping! """
        front = self.arr[1]
        self.swap( 1, -1 ) # back to front
        self.arr.pop() 
        # sink down the backman
        j = 1
        while True:
            c = self.maxchild( j )
            if c == -1: break
            if self.lt( j, c ): # parent < child?
                self.swap( j, c )
                j = c
            else: break
        return front

    def empty( self ):
        ''' Ck if empty. '''
        return len( self.arr ) == 1

def dijkstra( grid ):
    '''
    '''
    # mark rough terrains to avoid queueing them
    dist = np.full( grid.shape, -1, dtype=int )
    dist[grid == 1] = 1
    dist[0, 0] = 0
    # this is the output, starting from the last vertex
    # it will give backtrace of scv's path
    prev = np.full( grid.shape, None )
    # create a priority queue using distances as keys.
    heap = Heap( dist )
    # loop until all points are visited
    while not heap.empty():
        # pop up the front
        u = heap.pop()
        # ck East\South nbors
        for j in 0, 1:
            # copy to list so we can modify it
            v = list( u )
            # moo in the given direction
            v[j] += 1
            # ck if out of bounds
            if v[j] == dist.shape[j]: continue
            # back to normal
            v = tuple( v )
            # avoid rough terrain
            if grid[v] == 1: continue
            # update distance if necessary
            if dist[v] < dist[u] + grid[u]:
                dist[v] = dist[u] + grid[u]
                prev[v] = u
                heap.update( v )            
    return prev        

def backtrace( scv, prev, grid ):
    v = prev[-1, -1]
    s = ""
    samples = 0
    print(grid)
    while v:
        if grid[v] == 2:
            samples += 1
            s += '[{},{}], '.format( *v )
            grid[v] = 0
        else:
            s += '({},{}), '.format( *v )
        v = prev[v]
    if 0 < samples:
        print( "SCV: {}, Samples: {}".format( scv, samples ))
        print( "Backtrace:", s[:-2] )
    return samples

def main():
    """
    """
    # load grid array from input file
    grid = np.loadtxt( 'NPUT', dtype=int )
    # initialize scv unit counter
    scv = 0
    # run dijkstra algorithm until no moar rock samples
    while True:
        prev = dijkstra( grid )
        # ck if done
        scv += 1
        samples = backtrace( scv, prev, grid )
        if samples == 0: break

if __name__ == '__main__':
    main()

# log:

File: Mars.c
#include <stdio.h>
#include <string.h>
////////////////////////////////////////////////////////
//                                          some globals
// I/O sucks
enum { CLEAR, ROUGH, ROCKSAMPLE };
#define P 4 // width
#define Q 4 // height
typedef int terrain_t[Q][P];
terrain_t grid = {
  {0, 2, 0, 2},
  {2, 1, 0, 2},
  {0, 0, 2, 2},
  {1, 2, 0, 0},
};
terrain_t dist;
void dist_init( void )
{
  memset( dist, 0xff, sizeof dist ); // -1
  **dist = 0;
}
////////////////////////////////////////////////////////
void dump_terrain( terrain_t t, int w )
{
  static char buf[BUFSIZ];
  static char fmt[] = "%_d";
  fmt[1] = '0' + w;

  for ( int i = 0; i < Q; ++i ) {
    *buf = 0;
    for ( int j = 0; j < P; ++j ) {
      snprintf( buf + w*j, BUFSIZ, fmt, t[i][j] );
    }
    printf( "[%s]\n", buf );
  }
}
////////////////////////////////////////////////////////
typedef struct vtx {
  int coor[2];
} vtx_t;

char *vtx2str( vtx_t *v )
{
  static char buf[] = "(--,--)";

  snprintf( buf, sizeof buf, "(%2d,%2d)", v->coor[0], v->coor[1] );

  return buf;
}

vtx_t prev[Q][P];

void prev_init( void )
{
  memset( prev, 0xff, sizeof prev );
}

void dump_prev( void )
{
  static char buf[BUFSIZ];

  for ( int i = 0; i < Q; ++i ) {
    *buf = 0;
    for ( int j = 0; j < P; ++j ) {
      strcat( buf, vtx2str( &prev[i][j] ));
    }
    puts( buf );
  }
}
////////////////////////////////////////////////////////////////
// * Heap Section * //
int bp = 0;        // heap's back pointer
vtx_t heap[1+P*Q]; // base-1 max heap (key: dist)
void svap( int i, int j )
{
  static vtx_t v;
  v = heap[i]; heap[i] = heap[j]; heap[j] = v;
}
int get( terrain_t t, vtx_t *v )
{
  return t[v->coor[0]][v->coor[1]];
}
int lt( int i, int j )
{
  return get( dist, &heap[i] ) < get( dist, &heap[j] );
}
void bubble( int j )
{
  int p; // parent

  while ( p = j >> 1 ) {
    if ( lt( p, j )) {
      svap( p, j );
      j = p;
    } else break;
  }
}
void insert( int i, int j )
{
  heap[++bp] = (vtx_t) { i, j };
  bubble( bp );
}
void makeheap()
{
  bp = 0;
  for ( int i = 0; i < Q; ++i ) {
    for ( int j = 0; j < P; ++j ) {
      if ( grid[i][j] != ROUGH ) {
        insert( i, j );
      }
    }
  }
}
int maxchild( int p )
{
  int l = p << 1; // left child

  if ( l == bp ) {
    return l;
  } else if ( l > bp ) {
    return -1;
  } else {
    int r = l + 1;
    if ( lt( l, r )) {
      return r;
    } else {
      return l;
    }
  }
}
vtx_t *pop()
{
  svap( 1, bp-- );

  int p = 1, c;

  for (;;) {
    if (( c = maxchild( p )) == -1 || lt( c, p )) {
      break;
    }
    svap( p, c );
    p = c;
  }
  return &heap[bp+1];
}
int find( vtx_t *v )
{
  for ( int j = 1; j <= bp; ++j ) {
    vtx_t *u = &heap[j];
    for ( int i = 0; i <= 1; ++i ) {
      if ( u->coor[i] != v->coor[i] ) {
        goto skip;
      }
    }
    return j;
skip:;
  }
  return -1;
}
void update( vtx_t *v )
{
  int j = find( v );
  bubble( j );
}
void dumpheap()
{
  if ( bp == 0 ) puts( "empty heap" );
  int level = 1; // 1, 3, 7, 15, ...
  static char buf[BUFSIZ] = "";
  static char info[] = "(--,--)[**]";
  for ( int j = 1; j <= bp; ++j ) {
    vtx_t *v = &heap[j];
    snprintf( info, sizeof info, "%s[%d]", vtx2str( v ), get( dist, v ));
    strcat( buf, info );
    if ( j == level || j == bp ) {
      puts( buf );
      *buf = 0;
      level = ( level << 1 ) + 1;
    }
  }
}
// Proper Dijkstra's
typedef enum { SOUTH, EAST } dir_t;
int backtrace()
{
  static int scv = 1;
  int samples = 0;
  int i = Q-1;
  int j = P-1;
  for (;;) {
    vtx_t *p = &prev[i][j];
    i = p->coor[0];
    j = p->coor[1];
    if ( i == -1 ) break;
    if ( get( grid, p ) == ROCKSAMPLE ) {
      ++samples;
      grid[i][j] = CLEAR;
    }
  }
  if ( samples ) {
    printf( "scv %d: %d samples\n", scv++, samples );
  }
  return samples;
}
int dijkstra()
{
  static int shape[] = { Q, P };

  dist_init();
  prev_init();
  makeheap();
  dump_terrain( grid, 2 );
  while (bp > 0) {
    vtx_t *v = pop();
    for ( dir_t d = SOUTH; d <= EAST; ++d ) {
      vtx_t u = *v;
      ++u.coor[d];
      if (u.coor[d] == shape[d] || get( grid, &u ) == 1 ) {
        continue;
      }
      int shortcut = get( dist, v ) + get( grid, v ); 
      if ( get( dist, &u ) < shortcut ) {
        int i = u.coor[0];
        int j = u.coor[1];
        dist[i][j] = shortcut;
        prev[i][j] = *v;
        update( &u );
      }
    }
  }
  return backtrace();
}
int main( void )
{
  while (dijkstra())
    ;
}
// log:
